### PR TITLE
return error codes on failures

### PR DESCRIPTION
**Pull request by giladreti**
many rz_egg_* apis can fail but failures were previously ignored by rz-gg (or just logged). exit on failure, and attempt to detect more possible failures.

 

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**
many apis in rizin can fail. for example, if the input is not a valid c code `clang` will fail to compile it and `rz_egg_load_file` will return error, but `rz-gg` will continue anyways. also many cmdline options require their args to have a specific format, but corrently will just fail silently.
return an error in all such cases.

### CODE REVIEW

File: librz/egg/egg.c [hunk 1]
```diff
@@ -618,12 +618,16 @@ RZ_API void rz_egg_finalize(RzEgg *egg) {
 	}
 }
 
-RZ_API void rz_egg_pattern(RzEgg *egg, int size) {
+RZ_API int rz_egg_pattern(RzEgg *egg, int size) {
```
Comment (XVilka) [inline librz/egg/egg.c]: Make it bool

File: librz/include/rz_egg.h [hunk 2]
```diff
@@ -208,7 +208,7 @@ RZ_API int rz_egg_compile(RzEgg *egg);
 RZ_API int rz_egg_padding(RzEgg *egg, const char *pad);
 RZ_API bool rz_egg_assemble(RzEgg *egg);
 RZ_API bool rz_egg_assemble_asm(RzEgg *egg, char **asm_list);
-RZ_API void rz_egg_pattern(RzEgg *egg, int size);
+RZ_API int rz_egg_pattern(RzEgg *egg, int size);
```
Comment (XVilka) [inline librz/include/rz_egg.h]: Same

File: librz/main/rz-gg.c [hunk 3]
```diff
@@ -67,6 +67,8 @@ static void list(RzEgg *egg) {
 }
 
 static int create(const char *format, const char *arch, int bits, const ut8 *code, int codelen) {
```
Comment (XVilka) [inline librz/main/rz-gg.c]: Same, please make it bool then

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/main/rz-gg.c [hunk 4]
```diff
@@ -470,12 +473,18 @@ RZ_API int rz_main_rz_gg(int argc, const char **argv) {
...
+			eprintf("Cannot add padding\n");
+			goto fail;
+		}
 	}
 
 	// add pattern
 	if (pattern) {
-		rz_egg_pattern(egg, rz_num_math(NULL, pattern));
+		if (!rz_egg_pattern(egg, rz_num_math(NULL, pattern))) {
+			eprintf("Cannot add pattern\n");
```
Comment (wargio) [inline librz/main/rz-gg.c]: For errors, please use `RZ_LOG_ERROR("egg: cannot add pattern\n");`

**Review by wargio — CHANGES_REQUESTED** [review]
Please convert all the `eprintf` used for errors into `RZ_LOG_*`

**Review by XVilka — APPROVED** [review]

**Comment by wargio** [comment]
Please fix also the endianness issue that you have introduced previously:
https://app.travis-ci.com/github/rizinorg/rizin/jobs/600984447#L3668
```
[XX] db/tools/rz_gg rz-gg -k darwin -a x86 -b 64 simple_cmp.r

RZ_NOPLUGINS=1 /home/travis/build/rizinorg/rizin/install/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc '!rz-gg -k darwin -a x86 -b 64 bins/other/rz-gg/simple_cmp.r' =

-- stdout

--- expected

+++ actual

@@ -1,1 +1,1 @@

-554889e54881ec80000000c745106869210ac7451400000000c7451800000000488d45104889850800000048c7c70000000048c7c7010000004889bd0900000048c7c70000000048c7c7020000004889bd0a000000488b45085058483b450a7f3248c7c00400000050488b45085048c7c00100000050488b3c24488b742408488b54241048c7c0040000000f054883c418eb2248c7c00000000050488b3c2448c7c0010000000f054883c4084881c4800000005dc3ebdc

+554889e54881ec80000000c745100a216968c7451400000000c7451800000000488d45104889850800000048c7c70000000048c7c7010000004889bd0900000048c7c70000000048c7c7020000004889bd0a000000488b45085058483b450a7f3248c7c00400000050488b45085048c7c00100000050488b3c24488b742408488b54241048c7c0040000000f054883c418eb2248c7c00000000050488b3c2448c7c0010000000f054883c4084881c4800000005dc3ebdc

[XX] db/tools/rz_gg rz-gg -k darwin -a x86 -b 32 simple_cmp.r

RZ_NOPLUGINS=1 /home/travis/build/rizinorg/rizin/install/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc '!rz-gg -k darwin -a x86 -b 32 bins/other/rz-gg/simple_cmp.r' =

-- stdout

--- expected

+++ actual

@@ -1,1 +1,1 @@

-5589e581ec80000000c745086869210ac7450c00000000c74510000000008d4508898504000000bb00000000bb01000000899d05000000bb00000000bb02000000899d06000000ff7504583b45067f226a04ff75046a018b1c248b4c24048b542408b80400000050cd8083c40483c40ceb1b6a008b1c24b80100000050cd8083c40483c40481c4800000005dc3ebe3

+5589e581ec80000000c745080a216968c7450c00000000c74510000000008d4508898504000000bb00000000bb01000000899d05000000bb00000000bb02000000899d06000000ff7504583b45067f226a04ff75046a018b1c248b4c24048b542408b80400000050cd8083c40483c40ceb1b6a008b1c24b80100000050cd8083c40483c40481c4800000005dc3ebe3

```

**Review by wargio — APPROVED** [review]

**Comment by wargio** [comment]
Please do not forget to fix https://github.com/rizinorg/rizin/issues/3486

**Comment by giladreti** [comment]
done

**Pull request merged**